client/DESIGN.md

```diff
@@ -0,0 +1,292 @@
+# Treode Client Library Interface
+
+The two major components of the Treode client library interface are the [Cache](#Cache) and the [Transaction](#Transaction).  
+
+Caches store key-value pairs retrieved from the server, as well as the read and value timestamps associated with each pair that Treode provides.  Caches are responsible for maintaining a connection to the DB server.  Users may create multiple caches to work on disjoint sections of the database simultaneously.  
+
+Transactions allows users to locally perform a sequence of database operations and, when finished, optimistically commit the result to the server by writing through the cache.  Users can choose upon creating a Transaction which cache to associate it with.
+
+<a name="Cache"></a>
+# Cache
+
+We provide an overview of the Cache class below, including its primary internal data structure, its public methods, and the form of its HTTP requests and responses.
+
+- Parameters: 
+    1. server: string (required)
+    2. port: string (port=80)
+    3. max\_age: int (max\_age=None)
+    4. no\_cache: bool (no\_cache=False)
+- Public Methods: 
+    1. read(table: string, key: string, read\_timestamp=currentTime, max\_age=None, no\_cache=False): value (JSON)
+    2. write(table: string, op\_list: list): status (string), throws StaleException
+- Internal Objects: 
+    1. HTTP library for establishing/managing connection to server (in Python, urllib3)
+    2. sorted map for maintaining cache (in Python, [skiplist](https://github.com/sepeth/python-skiplist))
+- Exceptions: 
+    1. StaleException
+    2. StaleException(Value_time: int)
+
+## Cache Structure Description
+
+The cache internally maintains (1) a connection to the DB server and (2) a sorted key-value map of cache entries.  
+
+### Database Connection
+
+The cache should utilize an HTTP Request/Response library (e.g., in Python, urllib3) to establish a connection with the DB server and read and write values through its course of use.  We describe in the HTTP Message Formatting section the expected format for DB requests and responses.
+
+[Here are examples](https://urllib3.readthedocs.org/en/latest/) of using the urllib3 library in Python to open and use a connection.
+
+### Sorted Key-Value Map of Cache Entries
+
+The cache is a ("table:key" map) of (sorted value maps), which we explain below.
```

Comment:
Now that I realized the skiplist referenced lacks floor and ceiling lookup, a two level map might make sense. I'd first look for a proper skiplist that has floor and ceiling lookup. If you can find one, use that as a one level map. If you cannot find one, then two level maps make sense.

If you cannot find one, the two level map would be `table : key -> value-time -> value : read-time`. The first level will be a regular hash table (fast) and the second one will be an array/list. We presume the second level will be short enough that searching it linearly does not pose a significant problem.

In this design, making the second level sorted does not really impact anything. We need to find (floor), add, replace and remove items. If the array/list is sorted, all those ops are O(n). If the array/list is not sorted, add is O(1) and the others are O(n). See, not much point to sorting it. We presume find (floor) is the more frequent operation, and it's O(n) in both cases. The sorted case makes that O(n/2), but we expect n to be small, so it's probably not worth the effort to make it sorted.

I think you should write this doc primarily as if the cache is a one-level table using a skip-list. Include a subsection that discusses the cache may be implemented be a two level table.